### PR TITLE
Migrate type utilities for Cost Insights

### LINKED ISSUE #3134: cost insights page not loading
The cost insights page is not loading because we are getting currencies as undefined and we are trying to loop over the undefined using map method in CurrencySelect component.

## Expected Behavior

cost-insights-page-fix

## Current Behavior

cost-insights-page-bug

## Possible Solution

1. We could check for the presence of config.currencies and based on that we can render the `<CurrencySelect/>` component
2. We could use defaultCurrencies from types as fallback

## Steps to Reproduce

1. Run the project using `npm run dev`
2. From the side navigation select **Cost Insights**
3. You should see a white screen with error in console

## Context

I am looking to do open source contribution and I find this project very interesting

## Your Environment

- NodeJS Version (v12): v14.13.1
- Operating System and Version (e.g. Ubuntu 14.04): macOS Catalina 10.15.7
- Browser Information: Google Chrome Version 86.0.4240.111

### PR DESCRIPTION
**Pull request by ryanvazquez**
## Hey, I just made a Pull Request!

Fixes #3134
 
**Context**
~~By convention, the Cost Insights plugin defines certain utilities such as enums and functions in the type definition file of the type to which they most strongly map. For example, `Currency` and `CurrencyType` in `Currency.ts`, `getPageFilters` and `Filters` in `Filters.ts`, etc. However, this appears to now break the imports of these type utilities in other files, as these now return `undefined` upon import.~~

Moving the example client from `app` to `plugins/cost-insights/src/api` triggered several circular import cycles which indirectly cause a number of imports to fail. One such case was `defaultCurrencies`, which caused the initial state of `useConfig` to be initialized to undefined for `currencies` resulting in a run time error.

```ts
import { Currency, defaultCurrencies } from '../types';

const initialState = {
  currencies: defaultCurrencies // undefined
}
```

An example of a cycle:
```
../../plugins/cost-insights/src/hooks/index.ts ->
../../plugins/cost-insights/src/hooks/useConfig.tsx ->
../../plugins/cost-insights/src/types/index.ts -> 
../../plugins/cost-insights/src/types/Alert.tsx -> 
../../plugins/cost-insights/src/components/ProjectGrowthAlertCard/index.ts -> 
../../plugins/cost-insights/src/components/ProjectGrowthAlertCard/ProjectGrowthAlertCard.tsx -> 
../../plugins/cost-insights/src/components/ResourceGrowthBarChartLegend/index.ts -> 
../../plugins/cost-insights/src/components/ResourceGrowthBarChartLegend/ResourceGrowthBarChartLegend.tsx ->
../../plugins/cost-insights/src/components/CostGrowth/index.ts -> 
../../plugins/cost-insights/src/components/CostGrowth/CostGrowth.tsx -> 
../../plugins/cost-insights/src/hooks/index.ts
```

Type utils in `Alerts.tsx` exacerbate these cycles by referencing components, who in turn reference other types, components. etc.

To fix these cycles and to prevent similar kinds of import cycles in the future, the client is moved out of `/src/api` and into the root plugin folder. Additionally, we made a decision that types should only import from other types, not components, preventing this "crossover" effect from happening. Certain type utilities which could introduce these dependencies directly, or indirectly, will be moved to their own utility files.

Credit to @abhishekjakhar for surfacing this bug
Credit to @brendasukh for her detective work locating the circular dependency 

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/spotify/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)

### CODE REVIEW

File: plugins/cost-insights/src/utils/change.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+ * limitations under the License.
+ */
+
+import { Cost, ChangeStatistic, MetricData } from '../types';
+import { aggregationSort } from '../utils/sort';
+
+export const EngineerThreshold = 0.5;
+
+export enum ChangeThreshold {
+  upper = 0.05,
```
Comment (brendasukh) [inline plugins/cost-insights/src/utils/change.ts:23]: Do we want to keep these in the type file ?

File: plugins/cost-insights/src/components/ResourceGrowthBarChartLegend/ResourceGrowthBarChartLegend.tsx [hunk 2]
```diff
@@ -19,7 +19,7 @@ import { Box, useTheme } from '@material-ui/core';
 import { LegendItem } from '../LegendItem';
 import { CostGrowth } from '../CostGrowth';
 import { currencyFormatter } from '../../utils/formatters';
-import { ChangeStatistic, CostInsightsTheme, Duration } from '../../types';
+import { ChangeStatistic, Duration, CostInsightsTheme } from '../../types';
```
Comment (brendasukh) [inline plugins/cost-insights/src/components/ResourceGrowthBarChartLegend/ResourceGrowthBarChartLegend.tsx:22]: Does the order here matter ?

File: plugins/cost-insights/src/types/Duration.ts [hunk 3]
```diff
@@ -29,74 +26,3 @@ export enum Duration {
   P1M = 'P1M',
   P3M = 'P3M',
 }
-
-export const DEFAULT_DATE_FORMAT = 'YYYY-MM-DD';
```
Comment (brendasukh) [inline plugins/cost-insights/src/types/Duration.ts:33]: Is there a reason to move this out to utils ?

**Review by brendasukh — APPROVED** [review]
LGTM aside from some questions about some of the enums/consts moved to utils file.

File: plugins/cost-insights/src/types/Duration.ts (hunk 3 above)
Comment (ryanvazquez) [reply plugins/cost-insights/src/types/Duration.ts:33]: Prior impl made these non-public. I'll roll back these changes and cut a separate PR

**Review by timbonicus — APPROVED** [review]
Oof, nice sleuthing on this one @brendasukh and @ryanvazquez. I'm guilty of introducing a `.tsx` file in the `types` folder which should be a smell and contributed to this problem.

Thanks @abhishekjakhar for reporting what turned out to be a rather hairy bug 🙌  🥳

**Comment by freben** [comment]
@Rugvip for visibility - is this to do with our specific tooling, ie is it a bug that this ended up as undefined values in runtime instead of build time errors?

**Pull request merged**